pkg/commands/helpers/kubectl_auth.go

```diff
@@ -0,0 +1,270 @@
+*/
+
+package helpers
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"crypto/x509/pkix"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/spf13/cobra"
+	"k8s.io/client-go/util/homedir"
+	"k8s.io/klog"
+	"k8s.io/kops/cmd/kops/util"
+	"k8s.io/kops/pkg/commands/commandutils"
+	"k8s.io/kops/pkg/pki"
+	"k8s.io/kops/pkg/rbac"
+	"k8s.io/kops/upup/pkg/fi"
+	"k8s.io/kubectl/pkg/util/i18n"
+)
+
+var (
+	kubectlAuthShort = i18n.T(`Helper for kubectl authentication integration`)
+)
+
+// HelperKubectlAuthOptions holds the options for generating an authentication token
+type HelperKubectlAuthOptions struct {
+	// ClusterName is the name of the cluster we are targeting
+	ClusterName string
+
+	// Duration specifies the desired duration of the credential
+	Duration time.Duration
+
+	// APIVersion specifies the version of the client.authentication.k8s.io schema in use
+	APIVersion string
+}
+
+// InitDefaults populates the default values of options
+func (o *HelperKubectlAuthOptions) InitDefaults() {
+	o.Duration = 1 * time.Hour
+}
+
+// NewCmdHelperKubectlAuth builds a cobra command for the kubectl-auth command
+func NewCmdHelperKubectlAuth(f *util.Factory, out io.Writer) *cobra.Command {
+	options := &HelperKubectlAuthOptions{}
+	options.InitDefaults()
+
+	cmd := &cobra.Command{
+		Use:   "kubectl-auth",
+		Short: kubectlAuthShort,
+		Run: func(cmd *cobra.Command, args []string) {
+			ctx := context.TODO()
+
+			err := RunKubectlAuthHelper(ctx, f, out, options)
+			if err != nil {
+				commandutils.ExitWithError(err)
+			}
+		},
+	}
+
+	cmd.Flags().StringVar(&options.APIVersion, "api-version", options.APIVersion, "version of client.authentication.k8s.io schema in use")
+	cmd.Flags().StringVar(&options.ClusterName, "cluster", options.ClusterName, "cluster to target")
+	cmd.Flags().DurationVar(&options.Duration, "validity", options.Duration, "attempt to issue a credential with this validity")
+
+	return cmd
+}
+
+// RunKubectlAuthHelper implements the kubectl auth helper, which creates an authentication token
+func RunKubectlAuthHelper(ctx context.Context, f *util.Factory, out io.Writer, options *HelperKubectlAuthOptions) error {
+	if options.ClusterName == "" {
+		return fmt.Errorf("ClusterName is required")
+	}
+
+	execCredential := &ExecCredential{
+		Kind: "ExecCredential",
+	}
+
+	switch options.APIVersion {
+	case "":
+		return fmt.Errorf("api-version must be specified")
+	case "v1alpha1":
+		execCredential.APIVersion = "client.authentication.k8s.io/v1alpha1"
+	case "v1beta1":
+		execCredential.APIVersion = "client.authentication.k8s.io/v1beta1"
+
+	default:
+		return fmt.Errorf("api-version %q is not supported", options.APIVersion)
+	}
+
+	cacheFilePath := cacheFilePath(f.KopsStateStore(), options.ClusterName)
+	cached, err := loadCachedExecCredential(cacheFilePath)
+	if err != nil {
+		klog.Warningf("cached credential %q was not valid: %v", cacheFilePath, err)
```

Comment:
Not sure this and the wrong API version count as warnings.

Reply:
I believe this is when the credential is an error, not just expired, but you're right - going with Infof